tslearn/metrics/cysax.py

```diff
@@ -0,0 +1,150 @@
+STUFF_cysax = "cysax"
+
+import numpy as np
+from numba import njit, prange
+from sklearn.linear_model import LinearRegression
+
+__author__ = "Romain Tavenard romain.tavenard[at]univ-rennes2.fr"
+
+DTYPE = float
+DTYPE_INT = int
+
+
+@njit(parallel=True)
+def inv_transform_paa(dataset_paa, original_size):
+    n_ts = dataset_paa.shape[0]
+    sz = dataset_paa.shape[1]
+    d = dataset_paa.shape[2]
+    i = 0
+    t = 0
+    di = 0
+    t0 = 0
+    seg_sz = original_size // sz
+    dataset_out = np.zeros((n_ts, original_size, d))
+
+    for i in prange(n_ts):
+        for t in range(sz):
+            t0 = t * seg_sz
+            for di in range(d):
+                dataset_out[i, t0 : t0 + seg_sz, di] = dataset_paa[i, t, di]
+    return dataset_out
+
+
+@njit(parallel=True)
+def cydist_sax(sax1, sax2, breakpoints, original_size):
+    assert sax1.shape[0] == sax2.shape[0] and sax1.shape[1] == sax2.shape[1]
+    s = 0.0
+    sz = sax1.shape[0]
+    d = sax1.shape[1]
+    t = 0
+    di = 0
+    for t in prange(sz):
+        for di in range(d):
+            if np.abs(sax1[t, di] - sax2[t, di]) > 1:
+                max_symbol = max(sax1[t, di], sax2[t, di])
+                min_symbol = min(sax1[t, di], sax2[t, di])
+                s += (breakpoints[max_symbol - 1] - breakpoints[min_symbol]) ** 2
+    return np.sqrt(s * float(original_size) / sz)
+
+
+@njit(parallel=True)
+def inv_transform_sax(dataset_sax, breakpoints_middle_, original_size):
+    n_ts = dataset_sax.shape[0]
+    sz = dataset_sax.shape[1]
+    d = dataset_sax.shape[2]
+    i = 0
+    t = 0
+    di = 0
+    t0 = 0
+    seg_sz = original_size // sz
+    dataset_out = np.zeros((n_ts, original_size, d))
+
+    for i in prange(n_ts):
+        for t in range(sz):
+            t0 = t * seg_sz
+            for di in range(d):
+                dataset_out[i, t0 : t0 + seg_sz, di] = breakpoints_middle_[
+                    dataset_sax[i, t, di]
+                ]
```

Comment:
I think you can replace third for-loop with `dataset_out[.., ..., :] = ...`. (similar to another comment I provided for another part of code)

Reply:
According to the previous cython file cysax.pyx, the input breakpoints_middle_ is a  numpy ndarray of dimension 1.
The input dataset_sax is a numpy ndarray of dimension 3 of integers.
I am not sure that we can simplify this loop using `dataset_out[.., ..., :] = ...`.

Reply:
I think `dataset_sax[i, t, di]` is an integer value as it is used as index of `breakpoints_middle_`. Hence, we 
should be able to do:
```
dataset_out[i, t0 : t0 + seg_sz] = breakpoints_middle_[dataset_sax[i, t]]
```

So, `dataset_sax[i, t]` is an array containing integers that are being used as indices of `breakpoints_middle_`. And, then, we put that array into `dataset_out[i, t0 : t0 + seg_sz, :]`

Reply:
@YannCabanes have you tried this?